Add implicit import proposal (#500)
ghc-proposals/ghc-proposals: proposals/0000-implicit-import.rst

                 implements the described feature.
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/497>`_.
.. contents::

The goal of this proposal is to enable direct access to any available module without requiring an import statement.
This removes an unnecessary restriction, making the language simpler.

Motivation
----------
The main purpose is to be able to directly use a package.
This is particularly useful for single use functions, such as ``Servant.serve``.
For example, this propose change would make the following program valid
(assuming a cabal file with build depends and default extensions):

::

 module Main where

 import Servant.API (Get, JSON, (:>))

 type MyApi = "answer" :> Get '[JSON] Integer

 server :: Servant.Server MyApi
 server = pure 42

 main :: IO ()
 main = Network.Wai.Handler.Warp.run 8080 app
   where
     app :: Network.Wai.Application
     app = Servant.serve (Data.Proxy.Proxy @MyApi) server

The status quo is not satisfactory because it requires adding explicit imports:

::

 -- Without this proposal, these imports are required

>  import Data.Proxy

Reviewer: ```suggestion
 import qualified Data.Proxy
```
This should also be `qualified`, since the example uses `Data.Proxy.Proxy`, right?

Author: It's not necessary, but your suggestion looks better.

 import qualified Network.Wai
 import qualified Network.Wai.Handler.Warp
 import qualified Servant

Managing such imports has the following negative effects:

- It increases the file length.
- The author needs to jump to the top of the file each time a new module is used.
- It complicates code refactoring when moving a definition requires moving (or copying) its import.

Proposed Change Specification
-----------------------------
I propose an extension to enable implicit qualified imports.

When the ``ImplicitQualifiedImport`` language extension is enabled,
unknown fully qualified symbols are gracefully resolved early in the compilation pipeline by

> injecting implicit ``import qualified`` statement when necessary.

Reviewer: ```suggestion
injecting an implicit ``import qualified`` statement when necessary.
```

Examples
--------
Here are some example functions that are often used locally and
it would be useful to call them without having to add an import statement:

> - Control.Concurrent.threadDelay
> - Data.Char.isAlpha
> - Data.Foldable.traverse_
> - Data.Maybe.mapMaybe
> - Data.Set.fromList
> - Data.Text.pack
> - System.Environment.getArgs
> - Text.Printf.printf

Reviewer: ```suggestion
- ``Control.Concurrent.threadDelay``
- ``Data.Char.isAlpha``
- ``Data.Foldable.traverse_``
- ``Data.Maybe.mapMaybe``
- ``Data.Set.fromList``
- ``Data.Text.pack``
- ``System.Environment.getArgs``
- ``Text.Printf.printf``
```
This is code, so it should be in code blocks. This also prevents `Data.Foldable.traverse_` from being interpreted as a link.

Author: Thanks!

> - Text.Printf.printf

Reviewer: I rarely see `threadDelay`, `getArgs` and `printf` being used, so perhaps these aren't the best examples, although this may be subjective.

Author: I'd be happy to use a different list :) I've added `Debug.Trace.trace` in the last commit.

Effect and Interactions
-----------------------
The proposed change enables using any module without requiring an import statement.

Interactions with existing language or compiler features:

> - Hidden imports (using ``hidding``) must not be available implicitely.

Reviewer: ```suggestion
- Hidden imports (using ``hiding``) must not be available implicitly.
```

> - Hidden imports (using ``hidding``) must not be available implicitely.

Reviewer: What exactly does this mean? That with `import Data.Maybe hiding (mapMaybe)`, using `Data.Maybe.mapMaybe` should not be valid?

Author: If an import statement explicitly hides a definition, then this extension should not enable access to the definition, like in your example. Or perhaps this should only be the case for qualified imports, e.g. `import qualified Data.Maybe hiding (mapMaybe)`, then `Data.Maybe.mapMaybe` should not be valid.

- Modules available through multiple package will be disambiguated using the PackageImports extension.
- Only unknown fully qualified names will be affected.

Costs and Drawbacks
-------------------
TBD: estimate development and maintenance costs.

> This extension may improve the language's learnability for novice users by:

Reviewer: Given that beginners usually don't use any language extensions, I don't think this would really affect beginners (unless it is enabled by default, of course). That would change however if this extension became part of a future `GHC20XX` language.

Author: I can't think of a situation where this extension would not be useful, thus I'm hopeful it could be added to a future `GHC20XX` language.

This removes an unnecessary restriction, making the language simpler.

Motivation
----------
The main purpose is to be able to directly use a package.
This is particularly useful for single use functions, such as ``Servant.serve``.
For example, this propose change would make the following program valid
(assuming a cabal file with build depends and default extensions):

::

 module Main where

 import Servant.API (Get, JSON, (:>))

 type MyApi = "answer" :> Get '[JSON] Integer

 server :: Servant.Server MyApi
 server = pure 42

 main :: IO ()
 main = Network.Wai.Handler.Warp.run 8080 app
   where
     app :: Network.Wai.Application
     app = Servant.serve (Data.Proxy.Proxy @MyApi) server

The status quo is not satisfactory because it requires adding explicit imports:

::

 -- Without this proposal, these imports are required
 import qualified Data.Proxy
 import qualified Network.Wai
 import qualified Network.Wai.Handler.Warp
 import qualified Servant

Managing such imports has the following negative effects:

- It increases the file length.
- The author needs to jump to the top of the file each time a new module is used.

> - It complicates code refactoring when moving a definition requires moving (or copying) its import.

Reviewer 2: I'm sure you can add another argument: because unqualified is the default, user have the habit to introduce an unqualified import, which leads to import conflicts (now or in the future). Qualified locally is better (in my opinion) than unqualified globally, which is the default behavior.

Author: That is also my opinion, thank you for the suggestion. It's added in the last commit.

Proposed Change Specification
-----------------------------
I propose an extension to enable implicit qualified imports.

When the ``ImplicitQualifiedImport`` language extension is enabled,
unknown fully qualified symbols are gracefully resolved early in the compilation pipeline by
injecting an implicit ``import qualified`` statement when necessary.
The pipeline might need an extra pass to collect missing imports.

> GHCi already implements this feature thanks to the flag ``-fimplicit-import-qualified``,
> but the flag does not seem to work with GHC.
> Moreover the flag is not documented in ``ghc --show-options``, but it is accepted.
> Thus this proposal deprecates the ``-fimplicit-import-qualified`` flag in favor of the
> ``-XImplicitQualifiedImport`` modifier, so that the behavior is consistent between GHC and GHCi.

Reviewer: ```suggestion
GHCi already implements this feature via the ``-fimplicit-import-qualified`` flag,
but it does not seem to work with GHC.
Moreover the flag is not documented in ``ghc --show-options``, but it is accepted.
Therefore this proposal deprecates the ``-fimplicit-import-qualified`` flag in favor of
``-XImplicitQualifiedImport``, so that the behavior is consistent between GHC and GHCi.
```

Author: Thank you very much for the corrections!

Examples
--------
Here are some example functions that are often used locally and
it would be useful to call them without having to add an import statement:

- ``Control.Concurrent.threadDelay``
- ``Data.Char.isAlpha``
- ``Data.Foldable.traverse_``
- ``Data.Maybe.mapMaybe``
- ``Data.Set.fromList``
- ``Data.Text.pack``
- ``Debug.Trace.trace``
- ``System.Environment.getArgs``
- ``Text.Printf.printf``

Effect and Interactions
-----------------------
The proposed change enables using any module without requiring an import statement.

Interactions with existing language or compiler features:

> - Fully qualified import with hidden declarations are not affected: with ``import Data.Maybe hiding (mapMaybe)``, using ``Data.Maybe.mapMaybe`` should not be valid.

Reviewer 3: What about explicitly limited imports? If I write

```haskell
import Control.Monad (liftM2)
```

will `Control.Monad.ap` be in scope?

How about renamed imports? If I write

```haskell
import Control.Monad as M hiding (ap)
```

I would expect that `M.ap` would not be valid but that `Control.Monad.ap` would be. Do you concur?

- Modules available through multiple package will be disambiguated using the PackageImports extension.
- Only unknown fully qualified names will be affected.

Costs and Drawbacks
-------------------
TBD: estimate development and maintenance costs.

This extension may improve the language's learnability for novice users by:

- Reducing the length of code samples, and,
- Simplying new module usage, e.g. for one-off experiments and temporary annotations.

The main drawback is that the extension may reduce the readability of a module:

> its external requirements would no longer be explicitely listed in the import section.

Reviewer 3: ```suggestion
its external requirements would no longer be explicitly listed in the import section.
```

Reviewer 3: I think this is actually a pretty big deal, especially when trying to refactor code to change the module dependency order.

Reviewer 3: Another potential downside: if a module uses this feature, it's harder to swap out its dependencies for ones with similar interfaces but different implementations. This may not be a big deal in context.

Author: I agree, thank you for the suggestions, I'll add them to the next follow-up.
